### DISCUSSION TITLE
Avoiding `nix flake update` in stand-alone flake

### DISCUSSION THREAD

**Discussion by refaelsh** [nix-community/nixvim#2315] [Q&A] [answered]
Here are my dotfiles: https://github.com/refaelsh/dotfiles.
I have a Nixvim flake on my hard drive that I created like this: `nix flake init --template github:nix-community/nixvim`, the so-called stand-alone configuration.

The question is: when ever I change my stand-alone Nixvim flake and rebuild NixOS, the changes are not taking effect. I need to do `nix flake update` and then rebuild NixOS, and only then the changes will take effect.
I want to avoid running nix flake update because it will update the lock files of all the flakes and I might wait a while for the NixOS rebuilding to finish if there were many changes in various flakes. Also, I want to run nix flake update as less frequently as possible.

**Comment by MattSturgeon** [comment]
This is the intended way to use flakes. The entire point of a lockfile is that your downstream flake is not affected by changes in your upstream flake until it explicitly updates its lockfile.

You _can_ however bump specific inputs instead of updating all of them:
```
nix flake lock --update-input nixvim
```

Although, if you want the two projects to stay in sync, perhaps it makes more sense for them to be both combined into a single flake with multiple outputs?

**Reply by refaelsh** [reply]
> Although, if you want the two projects to stay in sync, perhaps it makes more sense for them to be both combined into a single flake with multiple outputs?

I don't understand that. Can you please elaborate?

**Reply by refaelsh** [reply]
> You can however bump specific inputs instead of updating all of them:

This is very clunky. I have numerous inputs to my NixOS flake.
What you are suggesting is, that every time I make a change in one of the inputs I need to think what input to update before rebuilding NixOS.

**Reply by khaneliman** [reply]
> > You can however bump specific inputs instead of updating all of them:
> 
> This is very clunky. I have numerous inputs to my NixOS flake. What you are suggesting is, that every time I make a change in one of the inputs I need to think what input to update before rebuilding NixOS.

It's normal dependency lockfile stuff. If you work with nugets, maven, npm, yarn etc. The idea is that you can bump all, one, or none of your dependencies and update and test stuff independently. When I want to just update my neovim i just do a `nix flake lock update khanelivim` or if I am waiting on a hyprland/nixpkgs/etc update.

**Reply by refaelsh** [reply]
> It's normal dependency lockfile stuff. If you work with nugets, maven, npm, yarn etc. The idea is that you can bump all, one, or none of your dependencies and update and test stuff independently.

Well, when you say it like this, I agree with you. It is indeed the same thing with npm and the likes. I get it now.

**Reply by MattSturgeon** [reply]
> > Although, if you want the two projects to stay in sync, perhaps it makes more sense for them to be both combined into a single flake with multiple outputs?
> 
> I don't understand that. Can you please elaborate?


TL;DR: To get a standalone nixvim, you don't need a standalone flake.

A flake can contain any number of "inputs" and "outputs".

You may find the following helpful:
- [flakes manual](https://nix.dev/manual/nix/2.24/command-ref/new-cli/nix3-flake.html#description)
- [flakes wiki page](https://wiki.nixos.org/wiki/Flakes)
- [Blog post](https://jade.fyi/blog/flakes-arent-real/)
- [nix.dev flakes overview](https://nix.dev/concepts/flakes.html)
- [The Nix Hour #4 \[comparing flakes to traditional Nix\]](https://www.youtube.com/watch?v=atmoYyBAhF4) (youtube)

**Reply by refaelsh** [reply]
> [Blog post](https://jade.fyi/blog/flakes-arent-real/)

Oh! This one looks interesting.

> [The Nix Hour #4 [comparing flakes to traditional Nix]](https://www.youtube.com/watch?v=atmoYyBAhF4)

And this one too.

All the rest I know and read.

**Reply by refaelsh** [reply]
> TL;DR: To get a standalone nixvim, you don't need a standalone flake.
> A flake can contain any number of "inputs" and "outputs".

A light bulb just went over my head. I think, I know what you mean, and I think I will be able to do it. But then I will not be able to tell a friend to just do `nix run github:refaelsh/dotfile?ref=nixvim` to get the same Neovim as I do.

**Reply by MattSturgeon** [reply, marked as answer]
> But then I will not be able to tell a friend to just do `nix run github:refaelsh/dotfile?ref=nixvim` to get the same Neovim as I do.

Assume you combine things into your top-level flake's outputs:

If you have many packages, and don't want your neovim to be your _default_ package, then you can tell a friend to do:
```
nix run github:refaelsh/dotfile#package-name

# e.g:
nix run github:refaelsh/dotfile#nvim

# or locally:
nix run .#nvim
```

If neovim _is_ what you would consider your "default" build, then you can alias your `packages.*.nvim` output as `packages.*. default` and the following will also work:

```
nix run github:refaelsh/dotfile

# which is shorthand for:
nix run github:refaelsh/dotfile#default
```

**Reply by MattSturgeon** [reply]
Essentially, `?ref` is saying "this flake is in a subdirectory" while `#foo` is saying "use the output `foo` on the specified flake".

You can be even more explicit, too, but it's usually unnecessary:

```
nix run github:refaelsh/dotfile#packages.x86_64-linux.nvim
```

**Reply by refaelsh** [reply]
Yep, that was the missing piece. I should have thought about `nix run github:refaelsh/dotfile#package-name` myself.

Thank you very much! :-)
